File: restaurant-management/python/services/restaurant_service.py
from models import SessionLocal, Table, Reservation

def find_optimal_table(guests_count: int, time: str) -> Table:
    """
    Wyszukuje pierwszy pasujący, najmniejszy wolny stolik dla danej liczby gości i godziny (podejście funkcyjne).
    """
    db = SessionLocal()
    try:
        tables = db.query(Table).all()
        reservations = db.query(Reservation).filter(Reservation.time == time).all()
        
        # Sprawdzamy zajętość stolika w podejściu funkcyjnym bez if/for
        is_reserved = lambda table_id: any(map(lambda r: r.table_id == table_id, reservations))
        
        # Filtrujemy wolne stoliki spełniające wymagania pojemności
        available = list(filter(lambda t: t.seats >= guests_count and not is_reserved(t.id), tables))
        
        # Zwracamy najmniejszy dopasowany stolik (Best-Fit) lub None
        return next(iter(sorted(available, key=lambda t: t.seats)), None)
    finally:
        db.close()


def book_table(user_id: str, table_id: int, time: str, guests_count: int) -> str:
    """
    Rezerwuje stolik (podejście funkcyjne bez if i pętli).
    """
    db = SessionLocal()
    try:
        table = db.query(Table).filter(Table.id == table_id).first()
        existing_table = db.query(Reservation).filter(Reservation.table_id == table_id, Reservation.time == time).first()
        existing_user = db.query(Reservation).filter(Reservation.user_id == user_id, Reservation.time == time).first()
        
        return "Table not found" if not table else \
               ("Table already reserved" if existing_table else \
                ("User already has a reservation at this time" if existing_user else \
                 ("Not enough seats" if table.seats < guests_count else \
                  _create_booking(db, user_id, table_id, time, guests_count))))
    finally:
        db.close()


def _create_booking(db, user_id, table_id, time, guests_count):
    db.add(Reservation(user_id=user_id, table_id=table_id, time=time, guests_count=guests_count))
    db.commit()
    return "Reservation successful"

File: restaurant-management/nodejs/services/restaurantService.js
const { Table, Reservation } = require('../models');

/**
 * Wyszukuje optymalny stolik dla określonej liczby gości i godziny (podejście funkcyjne).
 * Zwraca najmniejszy pasujący stolik, który nie jest w tym czasie zarezerwowany.
 */
async function findOptimalTable(guestsCount, time) {
  const tables = await Table.findAll();
  const reservations = await Reservation.findAll({ where: { time } });

  // Funkcyjne sprawdzenie czy stolik jest zajęty (bez pętli i ifów)
  const isReserved = (tableId) => reservations.some(r => r.tableId === tableId);

  // Filtrowanie stolików spełniających warunki pojemności i dostępności
  const available = tables.filter(t => t.seats >= guestsCount && !isReserved(t.id));

  // Sortowanie po liczbie miejsc (rosnąco) i wybór pierwszego optymalnego stolika (Best-Fit)
  return available.sort((a, b) => a.seats - b.seats)[0] || null;
}

/**
 * Dokonuje rezerwacji stolika (podejście funkcyjne bez instrukcji if).
 */
async function bookTable(userId, tableId, time, guestsCount) {
  const table = await Table.findByPk(tableId);
  const existing = await Reservation.findOne({ where: { tableId, time } });

  // Wyrażenia trójargumentowe z instrukcją przecinkową dla operacji zapisu bazy danych
  return !table 
    ? 'Table not found'
    : existing 
      ? 'Table already reserved'
      : table.seats < guestsCount 
        ? 'Not enough seats'
        : (await Reservation.create({ tableId, time, guestsCount, userId }), 'Reservation successful');
}

module.exports = { findOptimalTable, bookTable };
